### PR TITLE
SelectableText long press on iOS

### LINKED ISSUE #62658: SelectableText has no selection handle on iOS when long press, only appears after choosing selecting all text
## Steps to Reproduce

1. Create a new app with 
`flutter create myapp`
2. In the main.dart, replace text with selectabletext : 
```
SelectableText(
   'You have pushed the button this many'
),
```
3. Run on Android, long press on the sentence, you can select and there are handles. 
4. Run on iOS, long press on the sentence, you can hit select all, and now you have all the text selected and handles are visible ; but there just after long press on text there is no handles, so you can not easily select just a few words. 

**Expected results:** 
I expect to have handles on iOS to select the text I want. 

**Actual results:** 
On iOS there is no handle. I can only hit select all, then reduce the selection.

Related to  https://github.com/flutter/flutter/issues/59600

Hello @justinmc as you asked me in this message https://github.com/flutter/flutter/issues/59600#issuecomment-665842294 I create this new issue. I did not took new screenshots, as they would have been the same as in the issue #59600. Is it ok for you ? 

<details>
<summary> flutter doctor -v </summary>

```

[✓] Flutter (Channel master, 1.21.0-6.0.pre.96, on Mac OS X 10.15.5 19F101,
    locale fr-FR)
    • Flutter version 1.21.0-6.0.pre.96 at /Users/user921712/Downloads/flutter
    • Framework revision a4d570f0dd (il y a 2 jours), 2020-07-29 08:26:05 -0700
    • Engine revision a9910e409c
    • Dart version 2.10.0 (build 2.10.0-1.0.dev 24c7666def)

 
[!] Android toolchain - develop for Android devices (Android SDK version 26.0.1)
    • Android SDK at /usr/local/Caskroom/android-sdk/3859397,26.0.1
    ✗ Flutter requires Android SDK 28 and the Android BuildTools 28.0.3
      To update using sdkmanager, run:
        "/usr/local/Caskroom/android-sdk/3859397,26.0.1/tools/bin/sdkmanager"
        "platforms;android-28" "build-tools;28.0.3"
      or visit https://flutter.dev/docs/get-started/install/macos#android-setup
      for detailed instructions.
    ✗ Android licenses not accepted.  To resolve this, run: flutter doctor
      --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.1

[!] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[!] IntelliJ IDEA Community Edition (version 2019.3.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] VS Code (version 1.47.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (1 available)
    • macOS (desktop) • macos • darwin-x64 • Mac OS X 10.15.5 19F101


```

</details>

### PR DESCRIPTION
**Pull request by justinmc**
## Description

Previously, iOS did not allow selection with long press in SelectableText (only Android did). This PR enables it.

| Old | New |
| ----- | --- |
| old | new |

## Related Issues

Closes https://github.com/flutter/flutter/issues/62658

## Tests

I've updated existing tests to check the correct behavior.  There were many tests that were testing incorrect behavior (that longpressing a word just shows the menu without selecting anything).

### CODE REVIEW

**Comment by justinmc** [comment]
@chunhtai Am I misunderstanding how this should work? It seems like this behavior was deliberately removed from iOS along with lots of tests to confirm that long pressing in iOS doesn't allow selection.  Take a look at the code changes I made and let me know.

I left two TODOs in the code while I confirm that this is what we want to do.

**Comment by chunhtai** [comment]
I copied the code from the original textfield, I am not sure why it was like this either. My guess is that probable has something to do with precision pointer like mouse or trackpad. cc @gspencergoog @dkwingsmt 
If this is indeed the desire behavior we should probably update that? Furthermore, if we don't need two different version between CupertinoTextField and TextField/Selectable, we can update the parent method and don't need to override this method in the sub classes

**Comment by justinmc** [comment]
Another thing I've discovered while working on this is that edge scrolling doesn't seem to work: https://github.com/flutter/flutter/issues/64059

**Comment by justinmc** [comment]
@gspencergoog and @dkwingsmt are there any special concerns for desktop/mouse text selection, and will I get myself into trouble by making the changes in this PR?  Adding you as reviewers.

**Comment by gspencergoog** [comment]
I think as long as long press on a mouse doesn't select, it's probably fine. It should only happen from a touch long press.

**Comment by justinmc** [comment]
Ok got it.  I'll plan to make these changes and just make sure that things work as I think they should, including longpressing with a mouse.  I'll @ mention for review when I'm done.  Thanks!

**Comment by justinmc** [comment]
One more bug I noticed while working on this is that longpressing in a normal field on iOS doesn't seem to have the correct behavior either: https://github.com/flutter/flutter/issues/64190

File: packages/flutter/lib/src/material/selectable_text.dart [hunk 1]
```diff
@@ -72,25 +72,11 @@ class _SelectableTextSelectionGestureDetectorBuilder extends TextSelectionGestur
...
-            to: details.globalPosition,
-            cause: SelectionChangedCause.longPress,
-          );
-          break;
-      }
+      renderEditable.selectWordsInRange(
+        from: details.globalPosition - details.offsetFromOrigin,
+        to: details.globalPosition,
+        cause: SelectionChangedCause.longPress,
+      );
```
Comment (justinmc) [inline packages/flutter/lib/src/material/selectable_text.dart:79]: Some of this code could probably be deduplicated between here, TextField, CupertinoTextField, and TextSelectionGestureDetectorBuilder, but I think I'll have to fix https://github.com/flutter/flutter/issues/64190 first.  I'll make sure I clean up any deduplication when I fix that issue.

Comment (justinmc) [reply packages/flutter/lib/src/material/selectable_text.dart:79]: Same with onSingleLongTapStart below.

File: packages/flutter/test/widgets/selectable_text_test.dart [hunk 2]
```diff
@@ -3933,38 +3971,4 @@ void main() {
       TargetPlatform.windows,
     }),
   );
-
-  testWidgets('The handles show after pressing Select All (iOS and Mac)', (WidgetTester tester) async {
```
Comment (justinmc) [inline packages/flutter/test/widgets/selectable_text_test.dart:3937]: I believe there is now no way to make Select All appear for a SelectableText on iOS, so I've removed this test.

File: packages/flutter/test/widgets/selectable_text_test.dart [hunk 3]
```diff
@@ -2957,10 +2992,13 @@ void main() {
     expect(firstCharEndpoint.length, 1);
     // The first character is now offscreen to the left.
     expect(firstCharEndpoint[0].point.dx, moreOrLessEquals(-125, epsilon: 1));
-  }, variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS,  TargetPlatform.macOS }));
+  },
+    variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS,  TargetPlatform.macOS }),
+    skip: true, // https://github.com/flutter/flutter/issues/64059
```
Comment (justinmc) [inline packages/flutter/test/widgets/selectable_text_test.dart:2997]: I don't know why this test was passing in the first place, but it started failing when I made this change, despite the behavior it describes being broken either way.  I've created the issue https://github.com/flutter/flutter/issues/64059 to deal with fixing it and unskipping it.

File: packages/flutter/test/widgets/selectable_text_test.dart (hunk 2 above)
Comment (justinmc) [reply packages/flutter/test/widgets/selectable_text_test.dart:3937]: That seems consistent with native iOS in my experimentation.

**Comment by justinmc** [comment]
@gspencergoog @dkwingsmt @chunhtai Ready for final review.

Here is the long press behavior on each platform now:

| iOS | Android | Mac |
| --- | ----- | -------- |
| Word selected, Copy shows. | Word selected, Copy and Select All show. | Nothing |
| Screen Shot 2020-08-19 at 2 54 09 PM | Screen Shot 2020-08-19 at 2 54 06 PM | Screen Shot 2020-08-19 at 2 54 28 PM |

That seems consistent with native in all cases to me.

File: packages/flutter/lib/src/material/selectable_text.dart [hunk 4]
```diff
@@ -118,22 +104,8 @@ class _SelectableTextSelectionGestureDetectorBuilder extends TextSelectionGestur
...
-          break;
-        case TargetPlatform.android:
-        case TargetPlatform.fuchsia:
-        case TargetPlatform.linux:
-        case TargetPlatform.windows:
-          renderEditable.selectWord(cause: SelectionChangedCause.longPress);
-          Feedback.forLongPress(_state.context);
-          break;
-      }
+      renderEditable.selectWord(cause: SelectionChangedCause.longPress);
```
Comment (chunhtai) [inline packages/flutter/lib/src/material/selectable_text.dart:107]: How do we differentiate between precision devices? the mouse and stylus should not make a selection right?

File: packages/flutter/lib/src/material/selectable_text.dart (hunk 1 above)
Comment (chunhtai) [reply packages/flutter/lib/src/material/selectable_text.dart:79]: I just double check it seems like selectable text is correct, but there will be different for long press on textfield when it has keyboard focus vs when it does not

File: packages/flutter/lib/src/material/selectable_text.dart (hunk 4 above)
Comment (justinmc) [reply packages/flutter/lib/src/material/selectable_text.dart:107]: Hmm I'm not sure but that would make sense. @gspencergoog Didn't you have a keyboard and mouse setup for an Android device?  Is there different behavior for longpressing with the mouse vs. a finger?

File: packages/flutter/lib/src/material/selectable_text.dart (hunk 1 above)
Comment (justinmc) [reply packages/flutter/lib/src/material/selectable_text.dart:79]: @chunhtai Are you talking about https://github.com/flutter/flutter/issues/64190 specifically or this PR?

File: packages/flutter/lib/src/material/selectable_text.dart (hunk 4 above)
Comment (justinmc) [reply packages/flutter/lib/src/material/selectable_text.dart:107]: That continues to work even after this change.  The behavior on all 3 platforms is shown in https://github.com/flutter/flutter/pull/63994#issuecomment-676748652.

**Review by chunhtai — COMMENTED** [review]
LGTM, although I found a clipping issue that is parallel to this. I am using iphone XR

ezgif com-optimize

**Review by chunhtai — APPROVED** [review]

**Comment by justinmc** [comment]
@chunhtai Is there an issue for that clipping bug?  It's not something I've seen before I don't think.

**Comment by chunhtai** [comment]
I don't think so, it is also hard to reproduce. I can only reproduce it on my physical device.

**Pull request merged**